File: captainhcg/go/200-number-of-islands.go
func numIslands(grid [][]byte) int {
    cnt := 0
    for x:= 0; x<len(grid); x+= 1{
        for y:=0; y < len(grid[0]); y+=1 {
            if grid[x][y] == '1'{
                cnt += 1
                flip(x, y, grid)
            }
        }
    }
    return cnt
}

func flip(x, y int, grid [][]byte){
    if x < 0 || x >= len(grid) || y < 0 || y >= len(grid[0]){
        return
    }
    if grid[x][y] == '0'{
        return
    }
    grid[x][y] = '0'
    flip(x+1, y, grid)
    flip(x-1, y, grid)
    flip(x, y-1, grid)
    flip(x, y+1, grid)
}

File: bwu/bfs/200-number-of-islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        res = 0
        if len(grid) == 0 or len(grid[0]) == 0:
            return res
            
        m, n = len(grid), len(grid[0])
        for i in xrange(m):
            for j in xrange(n):
                if grid[i][j] == '1':
                    res += 1
                    self.bfs(grid, m, n, i, j)
        return res
        
    def bfs(self, grid, m, n, i, j):
        delta_x = [-1, 1, 0, 0]
        delta_y = [0, 0, -1, 1]
        queue = [(i, j)]
        grid[i][j] = '0'
        while queue:
            x, y = queue.pop(0)
            for dx, dy in zip(delta_x, delta_y):
                if 0 <= x + dx < m and 0 <= y + dy < n and grid[x + dx][y + dy] == '1':
                    queue.append((x + dx, y + dy))
                    grid[x + dx][y + dy] = '0'
